com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java

```diff
@@ -478,16 +468,20 @@ public Task<?> createTask(IResource inputResource) throws CompileExceptionError
      * @return task
      * @throws CompileExceptionError
      */
-    public Task<?> createTask(IResource inputResource, Class<? extends Builder<?>> builderClass) throws CompileExceptionError {
+    public Task createTask(IResource inputResource, Class<? extends Builder> builderClass) throws CompileExceptionError {
         // It's possible to build the same resource using different builders
         String key = inputResource.getPath()+" "+builderClass;
-        Task<?> task = tasks.get(key);
+        if (!circularDependencyChecker.add(key)) {
```

Comment:
A very neat solution! I like it.

Reply:
This solution became possible after refactoring and building the dependency tree during the creation stage. It made the implementation of such a detector really trivial.